README
dedup util
----------

dedup util is a lightweight open-source file packing tool. It is based on 
block-level data de-duplication technology, and can reduce the data capacity
to save storage space.

Internal data layout of dedup package as followed (in src/dedup.h):
deduplication file data layout
+---------------------------------------------------------------------+
|  header  |  logic block data  |  unique block data |  file metadata |
+---------------------------------------------------------------------+

file metedata entry layout
+---------------------------------------------------------------+
|  entry header  |  pathname  |  entry data  |  last block data |
+---------------------------------------------------------------+

entry data layout
+---------------------------------------+
| bid 1 | bid 2 | ... | bid n-1 | bid n |
+---------------------------------------+

Features
-------
1. append/remove files into/from package.

2. no MD5 collsion problem, but loss some performance.

3. support both variable-length and fixed-length segmentation. 


Source
------

project URL:    https://sourceforge.net/projects/deduputil
svn repository: https://deduputil.svn.sourceforge.net/svnroot/deduputil

Building
--------

1. checkout source code
  svn co https://deduputil.svn.sourceforge.net/svnroot/deduputil deduputil
 
2. install libz-dev
  apt-get install libz-dev
  if it does not work, please try other installations.

3. compile and install
  ./gen.sh (if need)
  ./configure
  make & make install

Command line
------------

Usage: dedup [OPTION...] [FILE]...
dedup tool packages files with deduplicaton technique.

Examples:
  dedup -c foobar.ded foo bar    # Create foobar.ded from files foo and bar.
  dedup -a foobar.ded foo1 bar1  # Append files foo1 and bar1 into foobar.ded.
  dedup -r foobar.ded foo1 bar1  # Remove files foo1 and bar1 from foobar.ded.
  dedup -t foobar.ded            # List all files in foobar.ded.
  dedup -x foobar.ded            # Extract all files from foobar.ded.
  dedup -s foobar.ded            # Show information about foobar.ded.

Options:
  -c, --creat      create a new archive
  -x, --extract    extrace files from an archive
  -a, --append     append files to an archive
  -r, --remove     remove files from an archive
  -t, --list       list files in an archive
  -s, --stat       show information about an archive
  -C, --chunk      chunk algorithms: FSP, CDC, SB, default is CDC
  -z, --compress   filter the archive through zlib compression
  -b, --block      block size for deduplication, default is 4096
  -H, --hashtable  hashtable backet number, default is 10240
  -d, --directory  change to directory, default is PWD
  -v, --verbose    print verbose messages
  -h, --help       give this help list

Platforms
---------

Linux only currently, maybe unix can work, and windows will be soon.

Author
------

Aigui Liu, focused on storage technologies, data mining and 
distributed computing.

Aigui Liu <aigui.liu@gmail.com> 2010.06.02


ChangeLog
v1.4.2 - 2011.08.24
  - add feature: symlink support, but follow the link

v1.4.1 - 2011.08.09
  - fix bug: md5 return empty string
  - fix bug: Cannot allocate memory when undedup package sometimes
  - fix bug: bus error on 64 bits machine sometimes

v1.4.0 - 2011.08.04
  - add feature: big file support

v1.3.1 - 2011.07.28
  - add hashdb and listdb: be ready to support big file

v1.3.0 - 2010.09.29
  - fix bug: file chunking error with sliding block

v1.3.0 - 2010.09.28
  - add feature: block level compression 

v1.2.3 - 2010.09.01
  - add cmd bsim: similarity detect between FILE1 and FILE2 based on block level
  - add cmd bencode: encode file into fingerprint lists based on chunks

v1.2.2 - 2010.08.24
  - add feature: wildcards support for package extraction and removing
  - add feature: file timestamps preservation

v1.2.1 - 2010.08.04
  - fix bug: file chunking error with sliding block 

v1.2.1 - 2010.07.28
  - fix bug: Bus error on some machine with 64 bits CPU

v1.2.0 - 2010.06.30
  - some fine tuning on v1.2

v1.2 - 2010.06.29
  - fix bug: dedup --stat does not work
  - code review and annotation

v1.2 - 2010.06.28
  - refactoring file chunking to imporve performance
  - refactoring package layout to avoid data copy
  - some fine tuning

v1.0 release - 2010.06.25
  - fix bug: can not run multiple instances simultaneously
  - clean temporary files when process terminates abnormally
  - fix bug: create package incorrectly when file exists
  - add adler rolling checksum for CDC and SB chunking
  - performance tuning on file chunking 

v1.0 beta - 2010.06.23
  - add stat information about an archive
  - add sliding block chunking, but performance is a big issue
  - fix bug: cmd long options can not be processed correctly

v1.0 alpha - 2010.06.22
  - fix bug: crashed when packing lots of files with CDC chunking
  - add --chunk command line option
  - add --hashfunc command line option

v1.0 alpha - 2010.06.21
  - variable-length segmentation
  - content-defined chunking
  - rabinhash32 fingerprinting

v1.0 alpha - 2010.06.17
  - refactoring to fit variable-length block chunking
  - fix some bugs

v1.0 alpha - 2010.06.17
  - release alpha version
  - fine tune program structure 
 
v1.0 - 2010.06.12
  - fix bug: crash when md5 collision happens

v1.0 - 2010.06.11
  - fix bug: same file can be saved to package more than one time

v1.0 - 2010.06.03
  - resolve md5 collsion problem, but lose some performance
  - fix remove bug: can only remove one file every time

v1.0 - 2010.06.01
  - merge codes into dedup_util 
  - fix bugs

v0.1 - 2010.05.26
  - add dedup_rm, remove file from deduped package

v0.1 - 2010.05.21
  - add dedup_add, add files to deduped package
  - extract specific file for undedup

v0.1 - 2010.05.20
  - add dedup_ls, list contents of deduped package 

v0.1 - 2010.01.07
  - keep mode of original file

v0.1 - 2010.01.03
  - parse argvs with getopt
  - rewrite Makefile with autoconf and automake 

v0.1 - 2009.12.27
  - package with deduplication feature
  - extract files from deduped package
  - z77 compress supported with zlib


src/bdiff.c
/* Copyright (C) 2010 Aigui Liu
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, visit the http://fsf.org website.
 */

#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <dirent.h>
#include <unistd.h>
#include <getopt.h>
#include <fcntl.h>
#include <dirent.h>
#include <errno.h>
#include "md5.h"
#include "hashtable.h"
#include "dedup.h"

/* block length */
static unsigned int g_block_size = BLOCK_SIZE;

/* hashtable backet number */
static unsigned int g_htab_backet_nr = BUCKET_SIZE;

/*
 * output block difference info to file or std.
 */
void output_bdiff(int fd, int block_no, int diff, off_t offset, unsigned int size, int debug)
{
	char buf[128] = {0};
	
	sprintf(buf, "#%d %d %ld %d\n", block_no, diff, offset, size);
	if (debug)
	{
		printf("%s", buf);
		return;
	}

	if (fd)
	{
		write(fd, buf, strlen(buf));
	}
	return;
}

/*
 * search block in dest file with block md5_checksum.
 * if matched, then offset += block_size, else offset++.
 */
off_t block_search(int fd, char *md5_checksum, off_t *offset, int debug)
{
	off_t boffset = -1;
	char *buf = NULL;
	unsigned char md5_checksum_dest[16 + 1] = {0};
	unsigned int rwsize = 0;

	if (-1 == *offset)
		return -1;

	buf = (char *)malloc(g_block_size);
	if (buf == NULL)
	{
		perror("malloc buf in block_search");
		return -1;
	}

	while (-1 != lseek(fd, *offset, SEEK_SET))
	{
		rwsize = read(fd, buf, g_block_size);
		/* get the tail or error */
		if (rwsize != g_block_size)
		{
			if (buf) free(buf);
			*offset = -1;
			return -1;
		}

		/* calculate md5 */
		md5(buf, rwsize, md5_checksum_dest);
		if (0 == strncmp(md5_checksum, md5_checksum_dest, 16))
		{
			boffset = *offset;
			*offset += g_block_size;
			break;
		}
		memset(buf, 0, g_block_size);
		memset(md5_checksum_dest, 0, 16 + 1);
		*offset++;
	}

	if (buf) free(buf);
	return boffset;
}

int bdiff(int fd_src, int fd_dest, int fd_tgt, hashtable *htab_inc, hashtable *htab_off, int debug)
{
	char *buf = NULL;
	unsigned int rwsize;
	off_t pos_src = -1, pos_dest = -1;
	off_t doffset = 0; /* offset in fd_dest */
	unsigned char md5_checksum[16 + 1] = {0};
	unsigned int block_no = 0;
	int ret = 0;

	buf = (char *)malloc(g_block_size);
	if (buf == NULL)
	{
		perror("malloc buf for block in bdiff");
		ret = errno;
		goto _BDIFF_EXIT;
	}

	pos_src = 0;
	while (rwsize = read(fd_src, buf, g_block_size)) 
	{
		/* if the last block */
		if (rwsize != g_block_size)
			break;

		/* calculate md5 */
		md5(buf, rwsize, md5_checksum);

		/* check hashtable with hashkey */
		off_t *b_offset = (block_id_t *)hash_value((void *)md5_checksum, htab_off);
		if (b_offset == NULL)
		{
			b_offset = (unsigned int *)malloc(BLOCK_ID_SIZE);
			if (NULL == b_offset)
			{
				perror("malloc b_offset in bdiff");
				ret = errno;
				break;
			}
			pos_dest = block_search(fd_dest, md5_checksum, &doffset, debug);

			/* insert hash entry and write unique block into bdata*/
			*b_offset = (pos_dest >= 0) ? pos_dest : pos_src;
			hash_insert((void *)strdup(md5_checksum), (void *)b_offset, htab_off);
			if (pos_dest >= 0)
			{
				unsigned int *bflag = (unsigned int *)malloc(sizeof(unsigned int));
				if (NULL == bflag)
				{
					perror("malloc bflag in bdiff");
					ret = errno;
					break;
				}
				hash_insert((void *)strdup(md5_checksum), (void *)bflag, htab_inc);
			}
			output_bdiff(fd_tgt, block_no, (pos_dest >= 0)?1:0, (off_t)*b_offset, g_block_size, debug);
		}
		else
		{
			unsigned int *b_inc = (block_id_t *)hash_value((void *)md5_checksum, htab_inc);
			output_bdiff(fd_tgt, block_no, (b_inc == NULL)? 0:1, (off_t)*b_offset, g_block_size, debug);
		}

		memset(buf, 0, g_block_size);
		memset(md5_checksum, 0, 16 + 1);
		pos_src += g_block_size;
		block_no++;
	}
	/* the last block */
	if (rwsize > 0 && rwsize != g_block_size)
	{
		output_bdiff(fd_tgt, block_no, 0, pos_src, rwsize, debug);
	}

_BDIFF_EXIT:
	if (buf) free(buf);

	return ret;
}

int bdiff_prepare(char *src_file, char *dest_file, char *tgt_file, int debug)
{
	int fd_src, fd_dest, fd_tgt = -1, ret = 0;
	/* record if block of source file is in dest file*/
	hashtable *htab_inc = NULL; 
	/* record block offset in source or dest file */
	hashtable *htab_off = NULL;

	if (-1 == (fd_src = open(src_file, O_RDONLY, 0755)))
	{
		perror("open source file");
		ret = errno;
		goto _BDIFF_PREPARE_EXIT;
	}

	if (-1 == (fd_dest = open(dest_file, O_RDONLY, 0755)))
	{
		perror("open dest file");
		ret = errno;
		goto _BDIFF_PREPARE_EXIT;
	}
	
	if(debug)
		goto _NO_OUTPUT_FILE;
	if (-1 == (fd_tgt = open(tgt_file, O_RDWR | O_CREAT, 0777)))
	{
		perror("open target file");
		ret = errno;
		goto _BDIFF_PREPARE_EXIT;
	}

_NO_OUTPUT_FILE:
	htab_inc = create_hashtable(g_htab_backet_nr);
	htab_off = create_hashtable(g_htab_backet_nr);
	if (NULL == htab_inc || NULL == htab_off)
	{
		perror("create_hashtable");
		ret = errno;
		goto _BDIFF_PREPARE_EXIT;
	}

	bdiff(fd_src, fd_dest, fd_tgt, htab_inc, htab_off, debug);

_BDIFF_PREPARE_EXIT:
	if (fd_src) close(fd_src);
	if (fd_dest) close(fd_dest);
	if (fd_tgt) close(fd_tgt);
	hash_free(htab_inc);
	hash_free(htab_off);
	
	return ret;
}

void usage()
{
	printf("Usage:  bdiff [OPTION...] <source file> <dest file> <output file>\n");
	printf("\ncompare files block by block.\n");
	printf("  -b, --block      block size for compare, default is 4096\n");
	printf("  -t, --hashtable  hashtable backet number, default is 10240\n");
	printf("  -d, --debug      print debug messages\n");
	printf("  -h, --help       give this help list\n");
	printf("\nReport bugs to <Aigui.Liu@gmail.com>.\n");
}

int main(int argc, char *argv[])
{
	int bhelp = 0, bdebug = 0;
	int ret = -1, c;
	struct option longopts[] =
	{
		{"block", 1, 0, 'b'},
		{"hashtable", 1, 0, 't'},
		{"debug", 0, &bdebug, 'd'},
		{"help", 0, &bhelp, 'h'},
		{0, 0, 0, 0}
	};

	/* parse options */
	while ((c = getopt_long (argc, argv, "b:t:dh", longopts, NULL)) != EOF)
	{
		switch(c) 
		{
		case 'b':
			g_block_size = atoi(optarg);
			break;
		case 't':
			g_htab_backet_nr = atoi(optarg);
			break;
		case 'd':
			bdebug = 1;
			break;
		case 'h':
		case '?':
		default:
			bhelp = 1;
			break;
		}
	}

	if (bhelp == 1 || (argc - optind) < 2)
	{
		usage();
		return 0;
	}

	if (bdebug)
		ret = bdiff_prepare(argv[optind], argv[optind + 1], NULL, bdebug);
	else
		ret = bdiff_prepare(argv[optind], argv[optind + 1], argv[optind + 2], bdebug);

	return ret;
}
